### PR TITLE
Fix build options configuration for Visual Studio projects

### LINKED ISSUE #79127: Inconsistent build options in a generated Visual Studio project
### Godot version

4.1dev [cdd2313]

### System information

Windows 11 22H2 - Visual Studio 2022 (17.6.4) - SCons 4.5.2 - Python 3.10.11

### Issue description

Multiple build options are ignored in a Visual Studio project generation step.

Build command line in Visual Studio project built with options `vsproj=yes dev_build=yes dev_mode=yes`:
Screenshot 2023-07-04 224622

Build command line in Visual Studio project built with options `vsproj=yes dev_build=yes verbose=yes warnings=extra tests=yes`:
Screenshot 2023-07-04 225646

As you can see in the second screenshot, options 'verbose' and 'warnings' are missing in the Visual Studio project. So is there any specific reason why only some options are transferred, or can I modify the `generate_vs_project` method to copy all original options (excluding `vsproj` of course) and create a PR? Also, I'm curious: why is the `progress` option explicitly set to `no` for Visual Studio projects?
https://github.com/godotengine/godot/blob/cdd2313ba27d0a2600a18e849b4c5d1fd6a6e351/methods.py#L851-L874

### Steps to reproduce

1. Build the engine with options `vsproj=yes dev_build=yes dev_mode=yes`
2. Open the generated project in Visual Studio
3. Open projects's NMake properties (Project->Properties->NMake)
4. Check `Build Command Line` and `Rebuild All Command Line` properties
5. Close Visual Studio
6. Rebuild the engine with options `vsproj=yes dev_build=yes verbose=yes warnings=extra tests=yes`
7. Open the generated project in Visual Studio
8. Open projects's NMake properties (Project->Properties->NMake)
9. Check `Build Command Line` and `Rebuild All Command Line` properties

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by CncealDVRflectN**
Fixes #79127

### CODE REVIEW

**Comment by anvilfolk** [comment]
This needs a rebase!

Also, I think special care needs to be had when using `scu_build`, though that is perhaps another problem for another PR, since that changes the files themselves as far as I can tell, and I had trouble with files not being in the solution when I used `scu_build` alongside `vsproj`. Perhaps it's easier to say in the documentation that those options aren't compatible?

**Comment by CncealDVRflectN** [comment]
I don't have a lot of knowledge of Godot and its build system, so I don't know about any specific flags that should be ignored specifically in Visual Studio projects, except those that I've filtered out in this PR. As I understand, Visual Studio just triggers SCons on build. If so, than I think we should pass as much user input as possible to VS projects, so Visual Studio Godot builds would be the same if we built them from terminal. 

Of course, there are exceptions. For instance, the `vsproj` flag itself. I've filtered it out in this PR as it's undesirable to regenerate a VS project on every build. 

In the end, I think it's better to exclude invalid options from user input for VS projects, than to just manually add a check for every new supported option to this VS specific script. This approach will make every new option available for VS projects by default, which is useful since there are a lot of contributors who don't use Visual Studio and who probably don't know that they should be adding their new options to this VS specific method. And as a result, it'll minimize the number of situations when contributors will get different builds using VS project compared to builds from terminal.

File: methods.py [hunk 1]
```diff
@@ -776,8 +776,18 @@ def add_to_vs_project(env, sources):
...
 
 
-def generate_vs_project(env, num_jobs, project_name="godot"):
+def generate_vs_project(env, original_args, project_name="godot"):
     batch_file = find_visual_c_batch_file(env)
+    filtered_args = original_args.copy()
+    # Ignore the "vsproj" option to not regenerate the VS project on every build
+    filtered_args.pop("vsproj", None)
+    # The "platform" option is ignored because only the Windows platform is currently supported for VS projects
+    filtered_args.pop("platform", None)
```
Comment (akien-mga) [inline methods.py:785]: This is fine for now, but we might want to look into supporting generating a VS project for other platforms.

Notably we should have a GDK port for Windows eventually, and console ports also typically require using Visual Studio.

**Review by akien-mga — APPROVED** [review]
Looks great! I was just ranting about why nobody was doing something like this, and why the `vsproj` option was still so limited with a list of hardcoded options being passed.

**Comment by akien-mga** [comment]
> I recently dug into this while looking into #81144, and that made me look into build options more carefully, and I'm not sure that we fully need all the options to be passed in. For example, `fast_unsafe` is already included by `dev_build`, and `tests` by `dev_mode`.

The `vsproj` config shouldn't need to do more logic than SCons itself does, since all that happens is just that VS calls scons when you press "Build" (the only added value is generating different target presets, and the boilerplate needed for VS to handle the source code as a project/solution).

So while specifying `scons dev_build=yes fast_unsafe=yes` is redundant, SCons already handles that. The latter will actually be used to override what `dev_build` would default to - so you can do `scons dev_build=yes fast_unsafe=no` and it will not use the `fast_unsafe` path. So the VS config should work the same and shouldn't filter out either option.

**Comment by anvilfolk** [comment]
Makes a lot of sense!

Just for future documentation purposes, `scu_build` is still broken with `vsproj`, since the VS solution includes the files generated by the SCUs and not the original ones IIRC, which leads to some really funky attempts at compilation. I believe the solution there will be to generate the vsproj before creating SCUs? TIWAGOS since I don't really understand SCUs 😅

**Comment by akien-mga** [comment]
Thanks!

**Comment by YuriSizov** [comment]
Cherry-picked for 4.1.2.

**Pull request merged**